src/shell/commands/executable.rs

```diff
@@ -98,13 +118,167 @@ impl ShellCommand for ExecutableCommand {
   }
 }
 
+enum CommandName {
+  Resolved(PathBuf),
+  Unresolved(UnresolvedCommandName),
+}
+
+struct ResolvedCommand<'a> {
+  command_name: CommandName,
+  args: Cow<'a, Vec<String>>,
+}
+
+#[derive(Error, Debug)]
+enum ResolveCommandError {
+  #[error(transparent)]
+  CommandPath(#[from] ResolveCommandPathError),
+  #[error(transparent)]
+  FailedShebang(#[from] FailedShebangError),
+}
+
+#[derive(Error, Debug)]
+enum FailedShebangError {
+  #[error(transparent)]
+  CommandPath(#[from] ResolveCommandPathError),
+  #[error(transparent)]
+  Any(#[from] anyhow::Error),
+}
+
+impl FailedShebangError {
+  pub fn exit_code(&self) -> i32 {
+    match self {
+      FailedShebangError::CommandPath(err) => err.exit_code(),
+      FailedShebangError::Any(_) => 1,
+    }
+  }
+}
+
+async fn resolve_command<'a>(
+  command_name: &UnresolvedCommandName,
+  context: &ShellCommandContext,
+  original_args: &'a Vec<String>,
+) -> Result<ResolvedCommand<'a>, ResolveCommandError> {
+  let command_path =
+    match resolve_command_path(command_name, &context.state, || {
+      Ok(std::env::current_exe()?)
+    }) {
+      Ok(command_path) => command_path,
+      Err(err) => return Err(err.into()),
+    };
+
+  // only bother checking for a shebang when the path has a slash
+  // in it because for global commands someone on Windows likely
+  // won't have a script with a shebang in it on Windows
+  if command_name.name.chars().any(|c| matches!(c, '\\' | '/')) {
```

Comment:
Oh I didn't know once can use backslash in shebags

Reply:
I don't think you can. I might as well remove that.